internal/warehouse/githubActions.go

```diff
@@ -0,0 +1,383 @@
+		w.logger.Info().Msgf("getting workflow runs and workflow jobs for workflow %s", *workflow.Name)
+		// we get a page of 30 workflow runs until next page is 0
+		for {
+
+			workflowRunsPage, response, err := w.githubClient.Actions.ListWorkflowRunsByID(ctx, owner, repo, *workflow.ID, opt)
+			if err != nil {
+				return err
+			}
+
+			resp = response
+
+			if *workflowRunsPage.TotalCount > 0 && len(workflowRunsPage.WorkflowRuns) > 0 {
+
+				if err := w.handleUpsertWorkflowRuns(ctx, workflowRunsPage.WorkflowRuns, repoID); err != nil {
+					return err
+				}
+
+				if err := w.handleWorkflowRunsJobs(ctx, owner, repo, repoID, workflowRunsPage.WorkflowRuns); err != nil {
+					return err
+				}
+			}
+
+			w.restRatelimitHandler(ctx, resp)
+
+			if response.NextPage == 0 {
+				break
+			}
+
+			opt.Page = resp.NextPage
+
+		}
+		opt.Page = 0
+
+	}
+
+	return nil
+}
+
+func (w *warehouse) handleWorkflowRunsJobs(ctx context.Context, owner, repo string, repoID uuid.UUID, workflowRunsPage []*github.WorkflowRun) error {
+
+	var resp *github.Response
+	opt := &github.ListWorkflowJobsOptions{
+		ListOptions: github.ListOptions{PerPage: 30},
+	}
+
+	for _, workflowRun := range workflowRunsPage {
+
+		// we get a page of 30 workflow run jobs until next page is 0
+		for {
+			workflowRunJobsPage, response, err := w.githubClient.Actions.ListWorkflowJobs(ctx, owner, repo, *workflowRun.ID, opt)
+
+			resp = response
+
+			if err != nil {
+				return err
+			}
+
+			if *workflowRunJobsPage.TotalCount > 0 && len(workflowRunJobsPage.Jobs) > 0 {
+				if err := w.handleWorkflowRunsJobLogs(ctx, owner, repo, repoID, workflowRunJobsPage.Jobs); err != nil {
+					return err
+				}
+			}
+
+			w.restRatelimitHandler(ctx, resp)
+
+			if response.NextPage == 0 {
+				break
+			}
+
+			opt.Page = resp.NextPage
+
+		}
+
+		// we reset the page for the next run
+		opt.Page = 0
+	}
+
+	return nil
+}
+
+func (w *warehouse) handleWorkflowRunsJobLogs(ctx context.Context, owner, repo string, repoID uuid.UUID, workflowRunJobsPage []*github.WorkflowJob) error {
+	var resp *github.Response
+	var log string
+
+	// we iterate over the workflowrunJobs page to get each log
+	for i, workflowJob := range workflowRunJobsPage {
+
+		workflowJobLog, response, err := w.githubClient.Actions.GetWorkflowJobLogs(ctx, owner, repo, *workflowJob.ID, true)
+
+		resp = response
+		if err != nil {
+			return err
+		}
+
+		if len(workflowJobLog.String()) > 0 {
+			log, err = w.parseJobLogs(workflowJobLog, strconv.Itoa(int(*workflowJob.ID)))
+			if err != nil {
+				return err
+			}
+		}
+
+		if err := os.RemoveAll(strconv.Itoa(i)); err != nil {
+			w.logger.Err(err)
+			return err
+		}
+
+		w.restRatelimitHandler(ctx, resp)
+
+		if err := w.handleUpsertWorkflowJobs(ctx, workflowJob, log, repoID); err != nil {
+			return err
+		}
+
+	}
+	return nil
+}
+
+func (w *warehouse) handleWorflowUpsert(ctx context.Context, workflows []*github.Workflow, repoID uuid.UUID) error {
```

Comment:
Name of function has a typo. It is also singular in name but accepts an array. I think based on our conversation the upserts should be singular for all three different upsert functions to keep the same pattern and should manage the looping in the function that calls the upsert.

Reply:
Done